clients/src/main/java/org/apache/kafka/common/utils/Utils.java

```diff
@@ -781,4 +781,29 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
         return Crc32.crc32(buffer.array(), buffer.arrayOffset() + start, size);
     }
 
+    /**
+     * Read data from the channel to the given byte buffer until there are no bytes remaining in the buffer or
+     * the channel has reached end of stream
+     * @param channel File channel containing the data to read from
+     * @param destinationBuffer The buffer into which bytes are to be transferred
+     * @param position The file position at which the transfer is to begin; must be non-negative
+     *
+     * @throws IllegalArgumentException If position is negative
+     * @throws IOException If reaching EOF or an I/O error occurs. See {@link FileChannel#read(ByteBuffer, long)}
+     */
+    public static void readFully(FileChannel channel, ByteBuffer destinationBuffer, long position) throws IOException {
+        if (position < 0) {
+            throw new IllegalArgumentException("The file channel position cannot be negative, but we got " + position);
+        }
+        long currentPosition = position;
+        int bytesRead;
+        do {
+            bytesRead = channel.read(destinationBuffer, currentPosition);
+            currentPosition += bytesRead;
+        } while (bytesRead != -1 && destinationBuffer.hasRemaining());
+        if (bytesRead == -1) {
+            throw new IOException(String.format("Reached EOF before filling up the buffer. Current file channel position = %d, buffer remaining = %d",
```

Comment:
Thanks for the feedback. Personally, I agree @junrao 's suggestion that we might need another read function to take care of the cases when reading EOF is expected. As for @ijuma 's "magic byte", I am not sure if I fully understand how it could be used, but I still think we have put too much responsibilities to readFully which otherwise should do only one thing: READING CHANNEL TO THE BUFFER. So I strongly recommend we refine the enclosing readInto to have all of these things pass the tests. 

What do you think?